File: solution/860. Lemonade Change/860.cpp
class Solution {
public:
    bool lemonadeChange(vector<int>& bills) {
        unordered_map<int, int> m;
        
        for (int x : bills) {
            if (x == 5) {
                m[x]++;
            } else if (x == 10) {
                if (m[5]-- < 0)
                    return false;
                m[10]++;
            } else if (x == 20) {
                if (m[10] > 0 && m[5] > 0) {
                    m[10]--;
                    m[5]--;
                } else if (m[5] >= 3) {
                    m[5] -= 3;
                } else {
                    return false;
                }
            }
        }
        return true;
    }
};

File: solution/860. Lemonade Change/860.py
class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        five = ten = 0
        for x in bills:
            if x == 5:
                five += 1
            elif x == 10:
                ten += 1
                five -= 1
            elif x == 20:
                if ten > 0:
                    ten -= 1
                    five -= 1
                else:
                    five -= 3
            if five < 0:
                return False
        return True
    
